code/modules/reagents/chemistry/machinery/chem_separator.dm

```diff
@@ -0,0 +1,239 @@
+			filling.color = mix_color_from_reagents(reagents.reagent_list)
+			. += filling
+	// Beaker overlay
+	if(beaker)
+		. += "[icon_state]_beaker"
+		// Beaker reagents overlay
+		if(beaker.reagents.total_volume)
+			var/threshold = null
+			for(var/i in 1 to fill_icon_thresholds.len)
+				if(ROUND_UP(100 * beaker.reagents.total_volume / beaker.reagents.maximum_volume) >= fill_icon_thresholds[i])
+					threshold = i
+			if(threshold)
+				var/fill_name = "[fill_icon_state]_b_[fill_icon_thresholds[threshold]]"
+				var/mutable_appearance/filling = mutable_appearance(fill_icon, fill_name)
+				filling.color = mix_color_from_reagents(beaker.reagents.reagent_list)
+				. += filling
+	// Thermometer overlay
+	var/threshold = null
+	for(var/i in 1 to temperature_icon_thresholds.len)
+		if(ROUND_UP(reagents.chem_temp - T0C) >= temperature_icon_thresholds[i])
+			threshold = i
+	if(threshold)
+		var/fill_name = "[icon_state]_temp_[temperature_icon_thresholds[threshold]]"
+		var/mutable_appearance/filling = mutable_appearance(icon_state, fill_name)
+		. += filling
+
+/obj/structure/chem_separator/proc/burn_attackby_check(obj/item/I, mob/living/user)
+	var/ignition_message = I.ignition_effect(src, user)
+	if(!ignition_message)
+		return
+	. = TRUE
+	user.visible_message(ignition_message)
+	fire_act(I.get_temperature())
+
+/obj/structure/chem_separator/attackby(obj/item/I, mob/user, params)
+	if(burn_attackby_check(I, user))
+		return
+	if(is_reagent_container(I) && !(I.item_flags & ABSTRACT) && I.is_open_container())
+		var/obj/item/reagent_containers/new_beaker = I
+		if(!user.transferItemToLoc(new_beaker, src))
+			return
+		replace_beaker(user, new_beaker)
+		to_chat(user, span_notice("You add [new_beaker] to [src]."))
+		update_appearance()
+		. = TRUE // no afterattack
+	else
+		return ..()
+
+/obj/structure/chem_separator/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
+	if(!user)
+		return FALSE
+	if(beaker)
+		if(burning)
+			stop()
+		if(!issilicon(user) && in_range(src, user))
+			user.put_in_hands(beaker)
+		else
+			beaker.forceMove(drop_location())
+		beaker = null
+	if(new_beaker)
+		beaker = new_beaker
+	update_appearance()
+	return TRUE
+
+/obj/structure/chem_separator/fire_act(exposed_temperature, exposed_volume)
+	if(!burning)
+		start()
+		return
+	. = ..()
+
+/obj/structure/chem_separator/extinguish()
+	if(burning)
+		stop()
+	. = ..()
+
+/obj/structure/chem_separator/proc/start()
+	if(!beaker)
+		return
+	if(beaker.reagents.total_volume >= beaker.reagents.maximum_volume)
+		return
+	if(!reagents.total_volume)
+		return
+	separating_reagent = reagents.reagent_list[1].type
+	burning = TRUE
+	update_appearance()
+	START_PROCESSING(SSobj, src)
+
+/obj/structure/chem_separator/proc/stop()
+	separating_reagent = null
+	burning = FALSE
+	update_appearance()
+	STOP_PROCESSING(SSobj, src)
+
+/obj/structure/chem_separator/proc/load()
+	if(burning)
```

Comment:
You might want to send a message to the user that you can’t fill it while using

Reply:
This proc is called from UI only, and the button is disabled there when they can't.